opentelemetry-auto-instrumentation/lib/opentelemetry-auto-instrumentation.rb

```diff
@@ -0,0 +1,149 @@
+    'resque' => 'OpenTelemetry::Instrumentation::Resque',
+    'ruby_kafka' => 'OpenTelemetry::Instrumentation::RubyKafka',
+    'sidekiq' => 'OpenTelemetry::Instrumentation::Sidekiq',
+    'sinatra' => 'OpenTelemetry::Instrumentation::Sinatra'
+  }.freeze
+
+  def detect_resource_from_env
+    env = ENV['OTEL_RUBY_RESOURCE_DETECTORS'].to_s
+    additional_resource = ::OpenTelemetry::SDK::Resources::Resource.create({})
+
+    env.split(',').each do |detector|
+      case detector
+      when 'container'
+        additional_resource = additional_resource.merge(::OpenTelemetry::Resource::Detector::Container.detect) if defined? ::OpenTelemetry::Resource::Detector::Container
+      when 'google_cloud_platform'
+        additional_resource = additional_resource.merge(::OpenTelemetry::Resource::Detector::GoogleCloudPlatform.detect) if defined? ::OpenTelemetry::Resource::Detector::GoogleCloudPlatform
+      when 'azure'
+        additional_resource = additional_resource.merge(::OpenTelemetry::Resource::Detector::Azure.detect) if defined? ::OpenTelemetry::Resource::Detector::Azure
+      end
+    end
+
+    additional_resource
+  end
+
+  def determine_enabled_instrumentation
+    env = ENV['OTEL_RUBY_ENABLED_INSTRUMENTATIONS'].to_s
+
+    env.split(',').map { |instrumentation| OTEL_INSTRUMENTATION_MAP[instrumentation] }
+  end
+
+  def require_otel
+    lib = File.expand_path('..', __dir__)
+    $LOAD_PATH.reject! { |path| path.include?('opentelemetry-auto-instrumentation') }
+    $LOAD_PATH.unshift(lib)
+
+    begin
+      required_instrumentation = determine_enabled_instrumentation
+
+      OpenTelemetry::SDK.configure do |c|
+        c.resource = detect_resource_from_env
+        if required_instrumentation.empty?
+          c.use_all # enables all instrumentation!
+        else
+          required_instrumentation.each do |instrumentation|
+            c.use instrumentation
+          end
+        end
+      end
+      OpenTelemetry.logger.info { 'Auto-instrumentation initialized' }
+    rescue StandardError => e
+      puts "Auto-instrumentation failed to initialize. Error: #{e.message}"
+    end
+  end
+end
+
+require 'bundler'
+
+container = ENV['OTEL_RUBY_RESOURCE_DETECTORS'].to_s.include?('container')
+google_cloud_platform = ENV['OTEL_RUBY_RESOURCE_DETECTORS'].to_s.include?('google_cloud_platform')
+azure = ENV['OTEL_RUBY_RESOURCE_DETECTORS'].to_s.include?('azure')
+
+# set OTEL_OPERATOR to false if not in autoinstrumentation-ruby image, default to /otel-auto-instrumentation-ruby
+# /otel-auto-instrumentation-ruby is set in opentelemetry-operator ruby.go
+operator_gem_path = ENV['OTEL_RUBY_OPERATOR'].nil? || ENV['OTEL_RUBY_OPERATOR'] == 'true' ? '/otel-auto-instrumentation-ruby' : nil
+additional_gem_path = operator_gem_path || ENV['OTEL_RUBY_ADDITIONAL_GEM_PATH'] || Gem.dir
+puts "Loading the additional gem path from #{additional_gem_path}" if ENV['OTEL_RUBY_AUTO_INSTRUMENTATION_DEBUG'] == 'true'
```

Comment:
Similar comment regarding diagnostic messages. Consider using `$stderr` for these so they can be siphoned by users that rely on structured logs coming from `$stdout`

Reply:
Updated to use $stdout.puts